allauth_2fa/adapter.py

```diff
@@ -13,12 +17,14 @@ def has_2fa_enabled(self, user):
         """Returns True if the user has 2FA configured."""
         return user_has_valid_totp_device(user)
 
-    def login(self, request, user):
+    def pre_login(self, request, user, **kwargs):
+        response = super().pre_login(request, user, **kwargs)
+        if response:
+            return response
+
         # Require two-factor authentication if it has been configured.
         if self.has_2fa_enabled(user):
-            # Cast to string for the case when this is not a JSON serializable
-            # object, e.g. a UUID.
-            request.session["allauth_2fa_user_id"] = str(user.id)
+            self.stash_pending_login(request, user, **kwargs)
```

Comment:
Since `stash_pending_login()` is private to `OTPAdapter`, it probably doesn't need to use `**kwargs` – maybe `login_kwargs=kwargs` here?